A72-open-telemetry-tracing.md

```diff
@@ -0,0 +1,428 @@
+A72: OpenTelemetry Tracing 
+----
+* Author(s): [Yifei Zhuang](https://github.com/YifeiZhuang)
+* Approver: [Eric Anderson](https://github.com/ejona86), [Mark
+  Roth](https://github.com/markdroth), [Doug Fawley](https://github.com/dfawley),
+[Feng Li](https://github.com/fengli79)
+* Status: Ready for Implementation
+* Implemented in: <language, ...>
+* Last updated: 2023-07
+* Discussion at: 
+
+## Abstract
+This proposal adds support for OpenTelemetry tracing and suggests migration
+paths away from OpenCensus tracing. Discussions include:
+* Context propagation between a gRPC client and server during migration.
+* Migrate a gRPC involved software binary that has both OpenTelemetry and 
+OpenCensus dependency.
+* The API surface to enable and configure OpenTelemetry tracing.
+
+Note that stats and logging are out of scope.
+
+## Background
+This work aligns with the community consensus to switch to OpenTelemetry as the 
+next generation OpenCensus. The latter is no longer maintained after July 31, 2023.
+
+Currently, gRPC supports OpenCensus based tracing in its grpc-census plugin, or 
+alike. gRPC trace is built with its core stream tracer and interceptor 
+infrastructures. A gRPC client when intercepting the call creates a child span
+with the parent span from the current context, and further creates attempt spans
+upon stream creation for each attempt. [gRFC A45][A45] describes cross language 
+design considerations about creating a child span for each individual call attempt. 
+A gRPC server uses span ID from the incoming request header as a parent span to 
+maintain the parent child span relationship with the gRPC client. To propagate 
+span context over the wire, gRPC uses metadata (header name: `grpc-trace-bin`) 
+and OpenCensus's binary format for (de)serialization. The header name is unique 
+from other census library propagators to differentiate with the application’s 
+tracing instrumentation.
+
+The following tracing information during an RPC lifecycle is captured:
+
+At the client, on parent span:
+* When the call is started, annotate name resolution completed if the RPC had 
+name resolution delay.
+* When the uncompressed size of some outbound data is revealed, annotate seq no.,
+type(Received) and wire message size.
+* When the call is closed, end the parent span with RPC status.
+
+On attempt span:
+* When span is created, add attribute `previous-rpc-attempts` that captures the 
+number of preceding attempts for the RPC, and attribute `transparent-retry` that 
+shows whether stream is a transparent retry.
+* When the stream is created on transport, annotate delayed load balancer pick 
+complete, if any.
+* When an outbound message has been sent, add a message event to capture seq no., 
```

Comment:
Interesting, it looks in Java, `MessageEvent` is used in `outboundMessageSent()` stream tracer methods for both client and server. Other places are using `annotation`.  
But, in OT the API is changed to only have `Event`, so it does not matter whether we used message event or annotation, everything will be an `Event`.
[A45 retry stats](https://github.com/grpc/proposal/blob/master/A45-retry-stats.md) does not mention that many details about cross language tracing information. I think it would be fine as long as languages keep generating the same information in OT tracing. Let me change the description of what is the current tracing information in more general way to avoid confusion.

Same for the wire size. Because java [StreamTracer](https://grpc.github.io/grpc-java/javadoc/io/grpc/StreamTracer.html) has those information, tracing just record it. So in C++ it what size information is captured in tracing?

Reply:
Isn't the `Event` just a string? We need to specify the contents of the string in that case, and I think the size should be the same as that of stats `Total bytes (compressed but not encrypted) sent across all request messages (metadata excluded) per RPC attempt; does not include grpc or transport framing bytes.`

Reply:
Also, I don't think we need "seq no.". I'm not even sure what it means.

Reply:
https://github.com/grpc/grpc-go/blob/master/stats/opencensus/trace.go#L107

We have something in our OpenCensus support that essentially annotates the message sent/recv events with an increasing number so we can see what order they were sent and received in. IIRC there was some async ordering issue where you could record these not in order so it is important to provide order. This was called out in the OpenCensus design, and I think this is what Ivy is referring to here.

Reply:
I don't think we have such an ordering issue in C++. Given that you just see it in trace UI in the right order, I don't see the benefit to mentioning the seq number, but it's simple enough that I can add it.

Reply:
I don't remember what the root reason was why this was included in the OpenCensus spec, but I'm sure it was called out for go tracing.

Reply:
1. Regarding seq.no, I believe Zach is right. From OpenCensus spec, a span records `MessageEvent` [span.addMessageEvent()](https://www.javadoc.io/doc/io.opencensus/opencensus-api/0.16.1/io/opencensus/trace/Span.html#addMessageEvent-io.opencensus.trace.MessageEvent-), and [messageId](https://opencensus.io/tracing/span/time_events/message_event/#id) identifies the message of the event. The purpose of the messageID is to 
 > The identifier for the message to help correlate between SENT and RECEIVED message events

Not that is not the http stream id, but "the sequential number of the message within the stream, starting from 0",  Java added it specifically for OpenCensus [#3461](https://github.com/grpc/grpc-java/pull/3461). It would be good that C++ can add that, but i'm fine to have it as is and add them later.

2. > Isn't the Event just a string? 

In OpenCensus, a span can record MessageEvent. It does not have a string name.
In OpenTelemetry, span can record Event. An Event has a string name and attributes (key/value pairs). I'll suggest a name update grfc, e.g. "Outbound message sent" and "Inbound message read". Java/Go also record both compressed and uncompressed message sizes when possible. The size is per message, and yes it does not include grpc framing. Mostly we just copy from the census implementation.

Reply:
Yeah, thinking about this, I think it was partly to handle the inordering between messages sent and received. A sequence number allows you to order between sent and received.

Reply:
I don't understand why we would want to put an order on sent and received messages. For example, if it is a bidi-stream, the application might not have enforced a specific ordering between the sent and received messages. gRPC *would* see them in *a* specific order but I still don't see the point of exposing that order to the users.

Reply:
Yash, I agree with you that the stream id doesn't seem to have critical values, maybe only for reference,  e.g. refer to message `1` is more convenient than referring to message `Thu, Jan 18, 2024, 11:39 AM GMT-8`. And also server side can match up with the client side message id.

The official description from OpenCensus for the use case of the sequence number:
> For example it can be useful when matching sequence/state numbers between protocol handshakes or for a streaming RPC. It is recommended to be unique within a span.

I don't see much point arguing the OpenCensus API for now. For Java/Go since users already have the sequence number in Census we just keep having it in OTel. For C++, no need to add if you don't already have it. And we move on. I'll add that.